server/src/main/java/org/elasticsearch/action/admin/indices/close/CloseIndexResponse.java

```diff
@@ -48,5 +77,225 @@ public void writeTo(StreamOutput out) throws IOException {
+        super.addCustomFields(builder, params);
+        builder.startObject("indices");
+        for (IndexResult index : indices) {
+            index.toXContent(builder, params);
+        }
+        builder.endObject();
+    }
+
+    @Override
+    public String toString() {
+        return Strings.toString(this);
+    }
+
+    public static class IndexResult implements Writeable, ToXContentFragment {
+
+        private final Index index;
+        private final @Nullable Exception exception;
+        private final @Nullable ShardResult[] shards;
+
+        public IndexResult(final Index index) {
+            this(index, null, null);
+        }
+
+        public IndexResult(final Index index, final Exception failure) {
+            this(index, Objects.requireNonNull(failure), null);
+        }
+
+        public IndexResult(final Index index, final ShardResult[] shards) {
+            this(index, null, Objects.requireNonNull(shards));
+        }
+
+        private IndexResult(final Index index, @Nullable final Exception exception, @Nullable final ShardResult[] shards) {
+            this.index = Objects.requireNonNull(index);
+            this.exception = exception;
+            this.shards = shards;
+        }
+
+        IndexResult(final StreamInput in) throws IOException {
+            this.index = new Index(in);
+            this.exception = in.readException();
+            this.shards = in.readOptionalArray(ShardResult::new, ShardResult[]::new);
+        }
+
+        @Override
+        public void writeTo(final StreamOutput out) throws IOException {
+            index.writeTo(out);
+            out.writeException(exception);
+            out.writeOptionalArray(shards);
+        }
+
+        public Index getIndex() {
+            return index;
+        }
+
+        public Exception getException() {
+            return exception;
+        }
+
+        public ShardResult[] getShards() {
+            return shards;
+        }
+
+        public boolean hasFailures() {
+            if (exception != null) {
+                return true;
+            }
+            if (shards != null) {
+                for (ShardResult shard : shards) {
+                    if (shard.hasFailures()) {
+                        return true;
+                    }
+                }
+            }
+            return false;
+        }
+
+        @Override
+        public XContentBuilder toXContent(final XContentBuilder builder, final Params params) throws IOException {
+            builder.startObject(index.getName());
+            {
+                if (hasFailures()) {
+                    builder.field("closed", false);
+                    if (exception != null) {
+                        builder.startObject("exception");
+                        ElasticsearchException.generateFailureXContent(builder, params, exception, true);
+                        builder.endObject();
+                    } else {
+                        builder.startObject("shards");
```

Comment:
Maybe we should call this failedShards instead, since it only contains the failed shards?

Reply:
I am fine with either way. I pushed https://github.com/elastic/elasticsearch/pull/39687/commits/a90aaeddeddb4053aad006bf19c89554b046f303.